### ISSUE TITLE
Why change \fsp behaviour on multi-codepoint graphemes?

### ISSUE THREAD

**Issue by amanosatosi** [libass/libass#697]
i'm using apps that one use new version of libass and one use old version and i wonder why devs removed this feature.

it happen when i use \fsp
this is old version from ubuntu
![image](https://github.com/libass/libass/assets/97550362/74b4d75f-76cf-4cfe-9b9e-ec396ae98e07)

and this is new version in aegisub
![失楽園    まふまふ_001_4855](https://github.com/libass/libass/assets/97550362/0f20b5f8-9c3b-429e-96d1-4b45a98f92e7)

for last en text, i used Myriad Pro and for burmese font, i used pyidaungsu. why it was removed in libass?

**Comment by TheOneric** [comment]
Is the “old version rom ubuntu” `<= 0.15.2`? If so, I strongly suspect the difference comes from the VSFilter-compatibility changes implemented in #441. As such they are intentional and the only right way to render real™ ASS since VSFilter is the authoritative reference. Existing ASS releases rely on exactly matching VSFilter.
If you can check [xy-VSFilter](https://github.com/pinterf/xy-VSFilter/releases), feel free to test if the new result actually matches and please report if it doesn't.

If you are authoring ASS subtitles for distribution, it is strongly recommended to ensure compatibility with xy-VSFilter since this is the only stable target. Unfortunately, this means you cannot just use `\fsp` for complex shaped or (most?) BiDi text. Instead you can edit the font or manually insert scaled spaces as necessary.
If you are doing on-the-fly conversions from non-ASS formats you can either set `Encoding=-1` in affected styles in the converter or enable `ASS_FEATURE_WHOLE_TEXT_LAYOUT` via libass API. However, note those extensions are both libass-specific and may be subtly tweaked in the future, i.e. unlike VSFilter-compatible ASS they do not guarantee pixel-perfect stable rendering in future versions.

**Comment by TheOneric** [comment]
Sorry, forgot to point out in the previous post:
If you do want to use `ASS_FEATURE_WHOLE_TEXT_LAYOUT` or `Encoding=-1`, bear in mind setting them affects more than just `\fsp` on multi-codepoint glyphs. Relevant documentation: [`WHOLE_TEXT_LAYOUT` (permalink to version of the time of writing)](https://github.com/libass/libass/blob/1a599b1db748dfc07df98cd37d8603edb87da2fd/libass/ass.h#L249-L267) and [`Encoding=-1`](https://github.com/libass/libass/wiki/Libass'-ASS-Extensions#encoding-1).

**Comment by amanosatosi** [comment]
> Is the “old version rom ubuntu” `<= 0.15.2`? If so, I strongly suspect the difference comes from the VSFilter-compatibility changes implemented in #441. As such they are intentional and the only right way to render real™ ASS since VSFilter is the authoritative reference. Existing ASS releases rely on exactly matching VSFilter and libass deviating from this in the past was a bug. If you can check [xy-VSFilter](https://github.com/pinterf/xy-VSFilter/releases), feel free to test if the new result actually matches and please report if it doesn't.
> 
> If you are authoring ASS subtitles for distribution, it is strongly recommended to ensure compatibility with xy-VSFilter since this is the only stable target. Unfortunately, this means you cannot just use `\fsp` for complex shaped or (most?) BiDi text. Instead you can edit the font or manually insert scaled spaces as necessary.
> 
> If you are doing on-the-fly conversions from non-ASS formats you can either set `Encoding=-1` in affected styles in the converter or enable `ASS_FEATURE_WHOLE_TEXT_LAYOUT` via libass API. However, note those extensions are both libass-specific and may be subtly tweaked in the future, i.e. unlike VSFilter-compatible ASS they do not guarantee pixel-perfect stable rendering in future versions. Also note, setting them affects _more_ than just `\fsp` on multi-codepoint glyphs. Relevant documentation: [`WHOLE_TEXT_LAYOUT` (permalink to version of the time of writing)](https://github.com/libass/libass/blob/1a599b1db748dfc07df98cd37d8603edb87da2fd/libass/ass.h#L249-L267) and [`Encoding=-1`](https://github.com/libass/libass/wiki/Libass'-ASS-Extensions#encoding-1).

it was not completely removed? how to turn it on? i'm not good at tech so much.

**Comment by TheOneric** [comment]
> it was not completely removed? how to turn it on?

Are you doing on-the-fly conversions from other subtitle formats or are rendering the output of such a converter in your program?
If yes, the two methods are already described in my previous post. To be able to give more specific advise, please describe exactly what your usecase is and how you use libass.

If you are not doing one of those two *and specifically if you are authoring native ASS files for distribution*, I strongly discourage using or relying on any of those methods. It will break rendering in standard renderers, at most work with libass and even players using libass might not work unless some non-standard configuration is applied.

**Comment by amanosatosi** [comment]
> Is the “old version rom ubuntu” `<= 0.15.2`? If so, I strongly suspect the difference comes from the VSFilter-compatibility changes implemented in #441. As such they are intentional and the only right way to render real™ ASS since VSFilter is the authoritative reference. Existing ASS releases rely on exactly matching VSFilter and libass deviating from this in the past was a bug. If you can check [xy-VSFilter](https://github.com/pinterf/xy-VSFilter/releases), feel free to test if the new result actually matches and please report if it doesn't.
> 
> If you are authoring ASS subtitles for distribution, it is strongly recommended to ensure compatibility with xy-VSFilter since this is the only stable target. Unfortunately, this means you cannot just use `\fsp` for complex shaped or (most?) BiDi text. Instead you can edit the font or manually insert scaled spaces as necessary.
> 
> If you are doing on-the-fly conversions from non-ASS formats you can either set `Encoding=-1` in affected styles in the converter or enable `ASS_FEATURE_WHOLE_TEXT_LAYOUT` via libass API. However, note those extensions are both libass-specific and may be subtly tweaked in the future, i.e. unlike VSFilter-compatible ASS they do not guarantee pixel-perfect stable rendering in future versions. Also note, setting them affects _more_ than just `\fsp` on multi-codepoint glyphs. Relevant documentation: [`WHOLE_TEXT_LAYOUT` (permalink to version of the time of writing)](https://github.com/libass/libass/blob/1a599b1db748dfc07df98cd37d8603edb87da2fd/libass/ass.h#L249-L267) and [`Encoding=-1`](https://github.com/libass/libass/wiki/Libass'-ASS-Extensions#encoding-1).



> > it was not completely removed? how to turn it on?
> 
> Are you doing on-the-fly conversions from other subtitle formats or are rendering the output of such a converter in your program? If yes, the two methods are already described in my previous post. To be able to give more specific advise, please describe exactly what your usecase is and how you use libass.
> 
> If you are not doing one of those two _and specifically if you are authoring native ASS files for distribution_, I strongly discourage using or relying on any of those methods. It will break rendering in standard renderers, at most work with libass and even players using libass might not work unless some non-standard configuration is applied.

i see. could you guide me how to build aegisub with that 0.15.2 version?

**Comment by TheOneric** [comment]
The original question of “why was it removed/changed” was already answered (changed for VSFilter incompatibility) and I linked docs on how to opt-in to VSF-incompatible modes and what consequences this has.

The revised question “why differs the result in Windows and Ubuntu” was already answered by yourself in your original post: you use different versions of libass in each OS.

I can't help you build and link other software, but plugging in an old version of libass isn't really a good solution and I again recommend to read the docs I linked before instead.
